### DISCUSSION TITLE
Quo component schemas with open vs. closed maps

### DISCUSSION THREAD

**Discussion by clauxx** [status-im/status-legacy#18842] [General]
We had a discussion a while back in the #_malli_ discord channel which we didn't finish. Best if we continue it in a more "centralized" place to collect our opinions and arrive at a decision that can go into our guidelines.

for **open** maps:
- allows more data to be passed to components than it uses i.e. usually from data coming from a subscription
- less rigid about prop requirements

for **closed** maps:
- ensures only the data needed by the component is passed on, avoiding unnecessary re-renders
- avoids bugs where props are misnamed or have typos

> _above is what I remember as the main points/arguments from the prior conversation_

**Comment by J-Son89** [comment]
My vote is for Quo components is that we used closed by default, and we allow open maps for previous components that would take a lot of refactoring to clean up immediately 👍 
It seems we generally want to avoid passing excess props along anyway so it's a better safeguard to have this imo

**Reply by J-Son89** [reply]
By Component schemas what does this mean exactly though, can we be more specific of where we want this?

**Reply by clauxx** [reply]
Quo components for now, as we dont write schemas for status-im components

**Comment by shivekkhurana** [comment]
**Note: Quo Components Only**

Utilizing closed maps implies foregoing the ability to drill props down (i.e., passing an object to children rather than explicitly passing keys).

Though it necessitates additional lines of code, **I fully endorse closed maps**. I advocate for their consistent use. It's acknowledged that implementing them mandates refactoring each component call when writing the schema.



![](https://sdk.bitmoji.com/render/panel/d8402ceb-7b29-43d2-9606-cc485fa83825-1348eaeb-87e6-4dfb-b64a-49a1b0e92955-v1.png?transparent=1&palette=1&width=160)

**Reply by cammellos** [reply]
I agree pretty strongly with this, the only exception is that we might not want to validate strictly what comes from status-go or the network, but we should pick and choose the stuff that we care about.

For example, say that we get from status-go:

`{"field1": 1, "field2": 2}`

if we validate the JSON RPC / signal returned value, it should not be strict, otherwise it would be really painful to upgrade.

But we should of course just pluck the field that we care about before anything else, and that can be put in a strict validation.

`(handle-something {:field-1 (.-field1 js-object) :field-2 (.-field2 js-object)})`  -> if `js-object` has `field3` it will not cause any schema failure, (we won't be using it until we explicitly write the code though, which I think it's fine)

I.e protect the boundaries of the application, but accept extra fields from status-go outside the boundaries and just ignore them, not sure if I expressed myself clearly

**Reply by shivekkhurana** [reply]
Plus 1. Backend stuff should be open. My comment above was only for Quo components.

**Reply by J-Son89** [reply]
@clauxx can we adjust the title to better reflect the discussion 👌

**Reply by ilmotta** [reply]
I don't want to over-generalize my position for closing or not schemas. Some in the Clojure community will be against closing them, others in favor. Malli was also born from the necessity to be able to close maps, something not possible with clojure.pec back then. To me, it's a case by case thing.

- Certain parts of the app would benefit from more strict instrumentation, like the Quo components. I totally agree with @shivekkhurana.
- Schemas for responses from status-go: I would _not close_ those schemas, otherwise we create maintainability issues in trying to keep different "type systems" synchronized. I fully agree with @cammellos
- Highly reused utility functions too (the general RPC `call` function, etc) those that are relatively stable, we have plenty of them, and I would close their map args whenever possible.
- I would add schemas and close map args (if needed) for macros because when macros fail the error message can be terrible. We don't have many macros, so this is not that important for us.

**Reply by ulisesmac** [reply]
It seems most (or all) of the team likes the idea of closed maps on components.

We already discussed it, but just wanted to share my thoughts here. I'd keep them **opened.**

My main reason is the flexibility, if we close them, I think the code will be filled with lots of `select-keys` or manually extracting keys to create a new map. It'd feel more like creating a type (like OOP) instead of passing data. e.g.

```clojure
(let [my-entity {,,,}]
  [component-1 (select-keys my-entity [:a :b :c])]
  [component-2 (select-keys my-entity [:a :c :e])]
  ;; or:
  [component-3 {:x (:x my-entity)
                :y (:y my-entity)
                :z (:z my-entity)}]
  [component-3 {:w (:w my-entity)
                :x (:x my-entity)}])
```
Too verbose for `component-3` with a few keys, and AFAIK `select-keys` is expensive to use.

In terms of dev experience I don't find any value on closed maps, I think a well defined opened map schema is able to catch missing or misspelled keys.

Also I imagine the code will evolve to create helper functions to create these "types" for components, since it'll be easier to have a single function that selects the specific keys for a component than remembering what are the keys, also it's less verbose, something like:

```clojure
(defn filter-props-for-component [props]
  (select-keys props [:foo :bar :x :y :z :w :a :b]))

(defn component []
  (let [my-entity {,,,}]
    [component-1 (filter-props-for-component my-entity)]
    ,,,
    [component-1 (filter-props-for-component my-entity)]
    ,,,
    [component-1 (filter-props-for-component my-entity)]))
```

And... do we really need/want this?

And about the argument of unnecessary re-renders when `my-entity` change, if those re-renders aren't impacting app perf or UX, I wouldn't mind, otherwise we can optimize by selecting the specific keys. Also let's keep in mind that these re-renders would always happen if we were using plain react, since it compares objects by reference.

We'll be dropping reagent, so let's see how these schemas play with the new implementation when components are no longer  Clojure functions, and unless we implement a mechanism to avoid re-renders, they will always happen, no matter if we pick specific keys.

**Reply by ilmotta** [reply]
I'm starting to be more in line with your arguments @ulisesmac. Just as we are trying to keep guidelines more flexible in the myriad of changes due to performance, we could do the same with schemas and start with the most flexible solution and learn from there.

**Reply by cammellos** [reply]
One thing though, I appreciate arguments both & cons, but

> learn from there

We did learn from this, we worked on this project long enough to know the cost of not being on top of performance, but just last week I had to change the community code because it was re-rendering the header every time we:

1) Checked permissions
2) Anything happened on a community (message received, spectated, etc)

This was happening because of two reasons:

1) Wrong structure of the database
2) Components were taking too many props as parameters

Both have been a factor, if either was coded correctly, we would not have had this issue, and it produced tangible work, since now have created extra issues to be picked up to correct the implementation. 
There's a difference between focusing on optimization (which we should not, unless necessary), and write code that is efficient, those are two different things, and the latter come at little to no cost most of the times.

With regards of open vs closed maps I think the way we should think about it is that:

"You must be this tall to use open maps."

Are we this tall? I think that's the question we want to focus on, and if we are, sure, but if we aren't, then it's either we need to get there somehow by being careful and building up experience, or we use guardrails (closed maps) until we are tall enough.

Or as Peter would put it:
![peter](https://github.com/status-im/status-mobile/assets/1017008/9120bf9f-4e1f-4454-8271-68c2b7884f60)

**Reply by J-Son89** [reply]
Poor Uncle Ben 🥲

**Reply by ilmotta** [reply]
In the clojure.spec days nobody was tall enough to close specs because it was not possible by design. But I like to have the choice to close in Malli. Whatever is decided will be fine by me. I won't object to any decision. But as @cammellos just pointed out, it looks like we're not tall enough yet.

The nicest part of it all to me is the discussion behind close vs open, even if the final decision can be somewhat arbitrary because both camps have merit 👍🏼

**Reply by ulisesmac** [reply]
@cammellos 
>  I had to change the community code because it was re-rendering the header every time
 ...
> This was happening because of two reasons:
>  1.  Wrong structure of the database
>  2. Components were taking too many props as parameters

Regarding to `2`,
Reagent re-executes its render functions when their props change and, AFAIK, that's an extra optimization on top of React's default behavior since Reagent just generates the VDOM needed and React the whole VDOM.

So, in the worst case, are we rendering the same as plain react (ofc, with the cost of using hiccup)?
As I said before, let's see if the performance arguments for using closed maps are still valid after dropping reagent.

In the meantime, I completely share what @ilmotta  said:

> I won't object to any decision. But as @cammellos just pointed out, it looks like we're not tall enough yet.
